Add ASTC HDR format variants

darksylinc opened the pull request:
Given issue #102696 and PR #102461 I figured we needed to properly support ASTC HDR.

I haven't tested it. I didn't touch the Metal backend, it does some weird stuff with lots of texture format setup I didn't have the time to decipher (MTLViewClass + calling addDataFormatDesc + calling addMTLPixelFormatDesc + setMTLPixFmtCapsIf).

darksylinc committed:
Add ASTC HDR format variants

stuartcarnie commented:
Metal already has mappings for the HDR variants, so you shouldn't have to do anything for the driver to support it:

https://github.com/godotengine/godot/blob/eef6f6adb1cdabf0bce7a18d05117afa951a6bc1/drivers/metal/pixel_formats.mm#L941-L955

This logic only disables it if the Apple GPU family is < and Apple6 GPU, which didn't support it.

stuartcarnie commented:
Ahh wait, I see, the mapping to the SFLOAT from Metal is missing; I can add that

BlueCube3310 commented:
Tested on Android with ARM Mali G76 MC4
Results:
| 4.4 Beta 3 (Mobile) | This PR (Mobile) | Compatibility |
|----------------------|-------------------|----------------|
| ![4_4_b3](https://github.com/user-attachments/assets/15df72a6-3faf-4fa7-a92c-f1d0c5935b5d) | ![pr](https://github.com/user-attachments/assets/f1306620-2d06-457b-89f7-aba7d92c14b6) | ![4_4_compat](https://github.com/user-attachments/assets/184fb8fe-c031-46ca-b539-a4cdbddcb1b6) |

darksylinc commented:
Oh! A different output! [PROGRESS](https://www.reddit.com/r/ProgrammerHumor/comments/tgogft/sometimes_progress_looks_like_failure/)!

@BlueCube3310 is there are a simple tutorial or youtube video I can follow and you can recommend to repro all this? I can look it up online but I'd waste time.

Basically the steps to repro so I can diagnose the problem (I coded this PR blindly).

> The new formats should probably be also added to RenderingDeviceCommons::get_compressed_image_format_block_byte_size and RenderingDeviceCommons::get_compressed_image_format_pixel_rshift (8x8 only), it might fix this

Thanks! I don't know how I missed those. I'll update the PR.

darksylinc commented:
>     The new formats should probably be also added to RenderingDeviceCommons::get_compressed_image_format_block_byte_size and RenderingDeviceCommons::get_compressed_image_format_pixel_rshift (8x8 only), it might fix this

OK I've updated the PR with these.

But this only raises more questions. For example `get_compressed_image_format_block_dimensions` and `get_compressed_image_format_block_byte_size` look like a code smell.

Even the code itself says things like:

````cpp
// Again, not sure about astc.
case DATA_FORMAT_ASTC_5x4_UNORM_BLOCK: // Unsupported
```

Which tells me this path is very untested. The way it does the math appears to be wrong for ASTC and would need refactoring. Fortunately I'm well versed in this area, but first I need to test ASTC actually works.

BlueCube3310 commented:
> Oh! A different output! [PROGRESS](https://www.reddit.com/r/ProgrammerHumor/comments/tgogft/sometimes_progress_looks_like_failure/)!
> 
> @BlueCube3310 is there are a simple tutorial or youtube video I can follow and you can recommend to repro all this? I can look it up online but I'd waste time.
> 
> Basically the steps to repro so I can diagnose the problem (I coded this PR blindly).
> 
> > The new formats should probably be also added to RenderingDeviceCommons::get_compressed_image_format_block_byte_size and RenderingDeviceCommons::get_compressed_image_format_pixel_rshift (8x8 only), it might fix this
> 
> Thanks! I don't know how I missed those. I'll update the PR.

For testing I downloaded the android editor build artifact and created an empty project. I then imported a random HDRi either .exr or .hdr, then set the compression mode to VRAM Compressed in the import settings. The preview in the inspector will show the results.

BlueCube3310 commented:
Tested it again, the latest version works correctly 
| Preview | Sky, energy 0.1 | Sky, energy 0.5 |
|----------|--------------------|--------------------|
| ![image](https://github.com/user-attachments/assets/251ad6e5-2067-4813-a065-b926f57fc639) | ![0_1_exp](https://github.com/user-attachments/assets/5d0f0fc2-b26f-45fd-ba72-9d04686d1148) | ![0_5_exp](https://github.com/user-attachments/assets/176d5f5f-b90a-4dcc-9f04-5d37e4cac349) |

BlueCube3310 reviewed (approved):
The code looks good to me, I've tested it on an android device and HDR textures now display correctly.

Additionally, devices that only support the base ASTC now properly fall back on the uncompressed counterparts of the HDR formats.

Repiteo commented:
Thanks!

akien-mga commented:
Cherry-picked for 4.4.1.
